resctrl/utils.go

```diff
@@ -0,0 +1,344 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Utilities.
+package resctrl
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fs2"
+	"github.com/opencontainers/runc/libcontainer/intelrdt"
+)
+
+const (
+	cpuCgroup             = "cpu"
+	rootContainer         = "/"
+	monitoringGroupDir    = "mon_groups"
+	processTask           = "task"
+	cpusFileName          = "cpus"
+	cpusListFileName      = "cpus_list"
+	schemataFileName      = "schemata"
+	tasksFileName         = "tasks"
+	modeFileName          = "mode"
+	sizeFileName          = "size"
+	infoDirName           = "info"
+	monDataDirName        = "mon_data"
+	monGroupsDirName      = "mon_groups"
+	noPidsPassedError     = "there are no pids passed"
+	noContainerNameError  = "there are no container name passed"
+	llcOccupancyFileName  = "llc_occupancy"
+	mbmLocalBytesFileName = "mbm_local_bytes"
+	mbmTotalBytesFileName = "mbm_total_bytes"
+	containerPrefix       = '/'
+	minContainerNameLen   = 2 // "/<container_name>" e.g. "/a"
+	unavailable           = "Unavailable"
+)
+
+var (
+	rootResctrl          = ""
+	pidsPath             = ""
+	processPath          = "/proc"
+	enabledMBM           = false
+	enabledCMT           = false
+	isResctrlInitialized = false
+)
+
+func Setup() error {
+	var err error
+	rootResctrl, err = intelrdt.GetIntelRdtPath(rootContainer)
+	if err != nil {
+		return fmt.Errorf("unable to initialize resctrl: %v", err)
+	}
+
+	if cgroups.IsCgroup2UnifiedMode() {
+		pidsPath = fs2.UnifiedMountpoint
+	} else {
+		pidsPath = filepath.Join(fs2.UnifiedMountpoint, cpuCgroup)
+	}
+
+	enabledMBM = intelrdt.IsMBMEnabled()
+	enabledCMT = intelrdt.IsCMTEnabled()
+
+	isResctrlInitialized = true
+
+	return nil
+}
+
+func prepareMonitoringGroup(containerName string, getContainerPids func() ([]string, error)) (string, error) {
+	if containerName == rootContainer {
+		return rootResctrl, nil
+	}
+
+	pids, err := getContainerPids()
+	if err != nil {
+		return "", err
+	}
+
+	if len(pids) == 0 {
+		return "", fmt.Errorf("couldn't obtain %q container pids, there is no pids in cgroup", containerName)
+	}
+
+	path, err := findControlGroup(pids)
+	if err != nil {
+		return "", fmt.Errorf("couldn't find control group matching %q container: %v", containerName, err)
+	}
+
+	if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
+		containerName = containerName[1:]
+	}
+
+	properContainerName := strings.Replace(containerName, "/", "-", -1)
+	monGroupPath := filepath.Join(path, monitoringGroupDir, properContainerName)
+
+	// Create new mon_group if not exists.
+	err = os.MkdirAll(monGroupPath, os.ModePerm)
```

Comment:
your are ignoring case, that someone created "own" (with "own naming scheme") mon_group inside existing ctrgroup - end results is:
- cadvsiro will accept exsting control groups but ignores existing mon_groups (if the don't fit your naming scheme)

It need to be explicitly documented  - or should behave more consistently (accept all - "passive mode" or ignore all "active/conflicting mode")